File: Matlab MRI/t1contrast.m
%{

This program plots the magnitudes after saturation
recovery and the resulting contrast
relative to gray matter.

%}

Mo = 1;
TR = 0:1:5000;

% get T1 values
prompt1 = 'CSF: ';
csfT1 = input(prompt1);

prompt2 = 'Gray matter: ';
grayT1 = input(prompt2);

prompt3 = 'White matter: ';
whiteT1 = input(prompt3);

prompt4 = 'Fat: ';
fatT1 = input(prompt4);


% equations for signal magnitude using given T1 values
CSF = Mo .* exp(-TR/csfT1);
gray = Mo .* exp(-TR/grayT1);
white = Mo .* exp(-TR/whiteT1);
fat = Mo .* exp(-TR/fatT1);

figure()


subplot(221)

% plot magnitudes
plot(TR, CSF);
hold on 
plot(TR, gray);
plot(TR, white);
plot(TR, fat);
hold off
grid on 
legend('CSF','Gray Matter', 'White Matter', 'Fat')

% plot contrast between gray matter and CSF
subplot(222)
csfcontrast = gray - CSF;
plot(TR, csfcontrast)
hold on
if csfcontrast<=0
    [M,I] = min(csfcontrast); 
else 
    [M,I] = max(csfcontrast) ;
end
plot(I, M, 'o');
hold off
title('Contrast Between Gray Matter and CSF')
grid on 

% plot contrast between gray matter and white matter
subplot(223)
whitecontrast = gray - white;
plot(TR, whitecontrast)
hold on
if whitecontrast<=0
    [M,I] = min(whitecontrast); 
else 
    [M,I] = max(whitecontrast) ;
end
plot(I, M, 'o');
hold off
title('Contrast Between Gray and White Matter')
grid on 

% plot contrast between gray matter and fat
subplot(224)
fatcontrast = gray - fat;
plot(TR, fatcontrast)
hold on
if fatcontrast<=0
    [M,I] = min(fatcontrast); 
else 
    [M,I] = max(fatcontrast) ;
end
plot(I, M, 'o');
hold off
title('Contrast Between Gray Matter and Fat')
grid on 

File: Python MRI/T1Contrast.py
"""

This program plots the magnitudes after saturation
recovery and the resulting contrast
relative to gray matter.

"""


import numpy as np
import matplotlib.pyplot as plt

Mo = 1
TR = np.arange(0,5000)

# get T1 values
csfT1 = int(input('T1 for CSF: '))
grayT1 = int(input('T1 for gray matter: ' ))
whiteT1 = int(input('T1 for white matter: '))
fatT1 = int(input('T1 for fat: '))

plt.figure()

# equations for signal magnitude using given T1 values
CSF = Mo * np.exp(-TR/csfT1)
gray = Mo * np.exp(-TR/grayT1)
white = Mo * np.exp(-TR/whiteT1)
fat = Mo * np.exp(-TR/fatT1)


# plot magnitudes
plt.subplot(221)
plt.plot(TR, CSF, label = 'CSF')
plt.plot(TR, gray, label = 'Gray Matter')
plt.plot(TR, white, label = 'White Matter')
plt.plot(TR, fat, label = 'Fat')
plt.legend()
plt.title('T1 Contrast')
plt.xlabel('TR (msecs)')
plt.ylabel('Magnitude')
plt.grid(True)
plt.xlim(0, 5000)
plt.ylim(ymin=0)

# plot contrast between gray matter and CSF
plt.subplot(222)
csfcontrast = gray - CSF
plt.plot(TR, csfcontrast)
if np.all(csfcontrast) <= 0:
    plt.plot(np.argmin(csfcontrast), np.min(csfcontrast), marker = 'o')
else:
    plt.plot(np.argmax(csfcontrast), np.max(csfcontrast), marker = 'o') 
plt.xlim(0, 5000)
plt.xlabel('TR (msecs)')
plt.ylabel('Contrast')
plt.grid(True)
plt.title('Contrast Between Gray Matter and CSF')

# plot contrast between gray matter and white matter
plt.subplot(223)
whitecontrast = gray - white
plt.plot(TR, whitecontrast)
if np.all(whitecontrast <= 0):
    plt.plot(np.argmin(whitecontrast), np.min(whitecontrast), marker = 'o')
else:
    plt.plot(np.argmax(whitecontrast), np.max(whitecontrast), marker = 'o')
plt.xlim(0, 5000)
plt.grid(True)
plt.xlabel('TR (msecs)')
plt.ylabel('Contrast')
plt.title('Contrast Between Gray and White Matter')

#plot contrast between gray matter and fat
plt.subplot(224)
fatcontrast = gray - fat
plt.plot(TR, fatcontrast)
if np.all(fatcontrast<=0):
    plt.plot(np.argmin(fatcontrast), np.min(fatcontrast), marker = 'o')
else:
    plt.plot(np.argmax(fatcontrast), np.max(fatcontrast), marker = 'o') 
plt.xlim(0, 5000)
plt.grid(True)
plt.xlabel('TR (msecs)')
plt.ylabel('Contrast')
plt.title('Contrast Between Gray Matter and Fat')


#show graphs
plt.tight_layout()
plt.show()
